annotation/equals_test.go

```diff
@@ -0,0 +1,301 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package annotation
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"go/types"
+	"reflect"
+	"strings"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/nilaway/util"
+	"golang.org/x/exp/maps"
+	"golang.org/x/tools/go/packages"
+)
+
+// This test file tests the implementation of the `equals` method defined for the interfaces `ConsumingAnnotationTrigger`,
+// `ProducingAnnotationTrigger` and `Key`.
+
+// Below are the helper utilities used in the tests, such as mock implementations of the interfaces and utility functions.
+
+// mockKey is a mock implementation of the Key interface
+type mockKey struct {
+	mock.Mock
+}
+
+func (m *mockKey) Lookup(m2 Map) (Val, bool) {
+	args := m.Called(m2)
+	return args.Get(0).(Val), args.Bool(1)
+}
+
+func (m *mockKey) Object() types.Object {
+	args := m.Called()
+	return args.Get(0).(types.Object)
+}
+
+func (m *mockKey) equals(other Key) bool {
+	args := m.Called(other)
+	return args.Bool(0)
+}
+
+// mockProducingAnnotationTrigger is a mock implementation of the ProducingAnnotationTrigger interface
+type mockProducingAnnotationTrigger struct {
+	mock.Mock
+}
+
+func (m *mockProducingAnnotationTrigger) CheckProduce(m2 Map) bool {
+	args := m.Called(m2)
+	return args.Bool(0)
+}
+
+func (m *mockProducingAnnotationTrigger) NeedsGuardMatch() bool {
+	args := m.Called()
+	return args.Bool(0)
+}
+
+func (m *mockProducingAnnotationTrigger) SetNeedsGuard(b bool) {
+	m.Called(b)
+}
+
+func (m *mockProducingAnnotationTrigger) Prestring() Prestring {
+	args := m.Called()
+	return args.Get(0).(Prestring)
+}
+
+func (m *mockProducingAnnotationTrigger) Kind() TriggerKind {
+	args := m.Called()
+	return args.Get(0).(TriggerKind)
+}
+
+func (m *mockProducingAnnotationTrigger) UnderlyingSite() Key {
+	args := m.Called()
+	return args.Get(0).(Key)
+}
+
+func (m *mockProducingAnnotationTrigger) equals(other ProducingAnnotationTrigger) bool {
+	args := m.Called(other)
+	return args.Bool(0)
+}
+
+// getImplementedMethods is a helper function that returns all the methods implemented by the struct "t"
+func getImplementedMethods(t *types.Named) []*types.Func {
+	visitedMethods := make(map[string]*types.Func) // helps in only storing the latest overridden implementation of a method
+	visitedStructs := make(map[*types.Struct]bool) // helps in avoiding infinite recursion if there is a cycle in the struct embedding
+	collectMethods(t, visitedMethods, visitedStructs)
+	return maps.Values(visitedMethods)
+}
```

Comment:
In this case, likely because we are parsing AST for files in a given package, it was not able to extract complete type information, particuarly, methods of embedded fields.